paasta_tools/cli/cmds/logs.py

```diff
@@ -549,6 +626,112 @@ def tail_logs(self, service, levels, components, clusters, raw_mode=False):
                 log.warn('Terminating.')
                 break
 
+    def print_logs_by_time(self, service, start_time, end_time, levels, components, clusters, raw_mode):
+        aggregated_logs = []
+
+        def callback(component, stream_info, scribe_env, cluster):
+            if stream_info.per_cluster:
+                stream_name = stream_info.stream_name_fn(service, cluster)
+            else:
+                stream_name = stream_info.stream_name_fn(service)
+
+            ctx = self.scribe_get_from_time(scribe_env, stream_name, start_time, end_time)
+            self.filter_and_aggregate_scribe_logs(ctx, scribe_env, stream_name, levels, service,
+                                                  components, clusters, aggregated_logs,
+                                                  filter_fn=stream_info.filter_fn,
+                                                  parser_fn=stream_info.parse_fn,
+                                                  start_time=start_time, end_time=end_time)
+
+        self.run_code_over_scribe_envs(clusters=clusters, components=components, callback=callback)
+
+        for line in aggregated_logs:
+            print_log(line, levels, raw_mode)
+
+    def print_last_n_logs(self, service, line_count, levels, components, clusters, raw_mode):
+        aggregated_logs = []
+
+        def callback(component, stream_info, scribe_env, cluster):
+            stream_info = self.get_stream_info(component)
+
+            if stream_info.per_cluster:
+                stream_name = stream_info.stream_name_fn(service, cluster)
+            else:
+                stream_name = stream_info.stream_name_fn(service)
+
+            ctx = self.scribe_get_last_n_lines(scribe_env, stream_name, line_count)
+            self.filter_and_aggregate_scribe_logs(ctx, scribe_env, stream_name, levels, service,
+                                                  components, clusters, aggregated_logs,
+                                                  filter_fn=stream_info.filter_fn,
+                                                  parser_fn=stream_info.parse_fn)
+
+        self.run_code_over_scribe_envs(clusters=clusters, components=components, callback=callback)
+        for line in aggregated_logs:
+            print_log(line, levels, raw_mode)
+
+    def filter_and_aggregate_scribe_logs(self, scribe_reader_ctx, scribe_env, stream_name,
+                                         levels, service, components, clusters,
+                                         aggregated_logs, parser_fn=None, filter_fn=None,
+                                         start_time=None, end_time=None):
+        # Scribe specific detail: all messages in app_output get logged at debug level, so we need to add that to
+        # the levels if its one of the components we're grabbing
+        if ('stdout' in components) or ('stderr' in components):
+            levels.append('debug')
```

Comment:
Does this mean that if we add the app_output component, we'll see all debug messages? Any way around that?